File: bindings/python/tests/test_url.py
"""Connection & URL routing (contract §1, §6)."""

import kevy
import pytest


def test_mem_anon_isolated():
    a = kevy.connect("mem://")
    b = kevy.connect("mem://")
    a.set("k", "1")
    assert b.get("k") is None  # two anonymous mem:// are independent
    a.close()
    b.close()


def test_mem_named_shares_store():
    name = "mem://shared-store-1"
    a = kevy.connect(name)
    b = kevy.connect(name)
    a.set("k", "v")
    assert b.get("k") == b"v"  # same backing store + bus
    a.close()
    b.close()


def test_file_shares_store(tmp_path):
    url = f"file://{tmp_path}/db"
    a = kevy.connect(url)
    b = kevy.connect(url)
    a.set("k", "v")
    assert b.get("k") == b"v"
    a.close()
    b.close()


def test_tls_rejected():
    for url in ("rediss://h", "kevys://h"):
        with pytest.raises(kevy.UnsupportedError):
            kevy.connect(url)


def test_auth_rejected():
    with pytest.raises(kevy.UnsupportedError):
        kevy.connect("redis://user:pass@host")


def test_unknown_scheme():
    with pytest.raises(kevy.InvalidInputError):
        kevy.connect("weird://host")


def test_file_empty_path():
    with pytest.raises(kevy.InvalidInputError):
        kevy.connect("file://")


def test_bad_db_index():
    with pytest.raises(kevy.InvalidInputError):
        kevy.connect("kevy://127.0.0.1:6379/notanint")


def test_is_embedded_flag():
    c = kevy.connect("mem://flag")
    assert c.is_embedded is True
    c.close()

File: bindings/cpp/test/test_url.cpp
// Connection & URL routing (contract §6, first block).
#include "backends.hpp"
#include "harness.hpp"
#include "kevy/client.hpp"

using namespace kevy;
using namespace kevy::test;

KEVY_TEST(url_mem_anon_isolated) {
  // Two anonymous mem:// are independent stores.
  Client a = Client::connect("mem://");
  Client b = Client::connect("mem://");
  a.set("k", "va");
  CHECK(!b.get("k").has_value());
  a.close();
  b.close();
}

KEVY_TEST(url_mem_named_shared) {
  // Same mem://<name> opened twice shares one store.
  std::string url = "mem://shared-" + std::to_string(42);
  Client a = Client::connect(url);
  Client b = Client::connect(url);
  a.set("k", "shared-val");
  auto v = b.get("k");
  CHECK(v.has_value());
  CHECK_EQ(*v, std::string("shared-val"));
}

KEVY_TEST(url_reject_tls) {
  CHECK_THROWS_KIND(Client::connect("rediss://localhost"), ErrorKind::Unsupported);
  CHECK_THROWS_KIND(Client::connect("kevys://localhost"), ErrorKind::Unsupported);
}

KEVY_TEST(url_reject_auth) {
  CHECK_THROWS_KIND(Client::connect("redis://user:pass@localhost"), ErrorKind::Unsupported);
}

KEVY_TEST(url_reject_unknown_scheme) {
  CHECK_THROWS_KIND(Client::connect("wat://localhost"), ErrorKind::InvalidInput);
  CHECK_THROWS_KIND(Client::connect("no-scheme-here"), ErrorKind::InvalidInput);
}

KEVY_TEST(url_reject_empty_file_path) {
  CHECK_THROWS_KIND(Client::connect("file://"), ErrorKind::InvalidInput);
}

KEVY_TEST(url_is_embedded_flag) {
  Client e = Client::connect(unique_mem_url());
  CHECK(e.is_embedded());
  if (server_available()) {
    auto s = spawn_server();
    Client r = Client::connect(s->url());
    CHECK(!r.is_embedded());
  }
}

KEVY_TEST(url_remote_db_select) {
  if (!server_available()) return;
  auto s = spawn_server();
  // kevy://…/0 does a SELECT 0 round-trip; tcp:// does not. Both must connect.
  Client a = Client::connect("kevy://127.0.0.1:" + std::to_string(s->port) + "/0");
  a.ping();
  Client b = Client::connect("tcp://127.0.0.1:" + std::to_string(s->port));
  b.ping();
}
